les/vflux/server/balance.go

```diff
@@ -256,16 +256,34 @@ func (n *NodeBalance) Priority(now mclock.AbsTime, capacity uint64) int64 {
 // in the current session.
 // If update is true then a priority callback is added that turns UpdateFlag on and off
 // in case the priority goes below the estimated minimum.
-func (n *NodeBalance) EstMinPriority(at mclock.AbsTime, capacity uint64, update bool) int64 {
+func (n *NodeBalance) EstimatePriority(now mclock.AbsTime, capacity uint64, addBalance int64, future, bias time.Duration, update bool) int64 {
 	n.lock.Lock()
 	defer n.lock.Unlock()
 
-	var avgReqCost float64
-	dt := time.Duration(n.lastUpdate - n.initTime)
-	if dt > time.Second {
-		avgReqCost = float64(n.sumReqCost) * 2 / float64(dt)
+	n.updateBalance(now)
+	b := n.balance
+	if addBalance != 0 {
+		offset := n.bt.posExp.LogOffset(now)
+		old := n.balance.pos.Value(offset)
+		if addBalance > 0 && (old > maxBalance-uint64(addBalance)) {
+			b.pos = utils.ExpiredValue{}
+			b.pos.Add(maxBalance, offset)
+		} else {
+			b.pos.Add(addBalance, offset)
+		}
 	}
-	pri := n.balanceToPriority(n.reducedBalance(at, capacity, avgReqCost), capacity)
+	if future > 0 {
+		var avgReqCost float64
+		dt := time.Duration(n.lastUpdate - n.initTime)
+		if dt > time.Second {
+			avgReqCost = float64(n.sumReqCost) * 2 / float64(dt)
+		}
+		b = n.reducedBalance(b, now, future, capacity, avgReqCost)
```

Comment:
AFAIK, we can merge `future` + `bias` together. 

e.g. 

```go

diff := 0
if future > 0 { diff +=  future}
if bias > 0 {diff += bias}
b := n.reducedBalance(b, now, diff, capacity, avgReqCost)
```

Reply:
No, the future estimation also applies an average req cost estimate while for the bias time we don't apply that.